examples/puppetmaster/test.sh

```diff
@@ -316,4 +379,107 @@ $ssl_certificate
+  echo "API key for $node_name: $api_key"
+
+  revoke_cert_for "$hostname"
+
+  set -x
+  run_with_docker_windows docker run --rm -t \
+    --hostname "${hostname}" \
+    puppet-agent \
+      powershell -Command "
+        # Allow resolution of the hostnames for conjur and puppet
+        Add-Content -Path 'c:\Windows\System32\Drivers\etc\hosts' -Value '${MAIN_HOST_IP} conjur.cyberark.com'
+        Add-Content -Path 'c:\Windows\System32\Drivers\etc\hosts' -Value '${MAIN_HOST_IP} puppet'
+        Add-Content -Path 'c:\conjur-ca.crt' -Value '$(cat $PWD/https_config/ca.crt)'
+
+
+        # Set conjur.conf equivalent with connection details
+        reg ADD HKLM\Software\CyberArk\Conjur /v ApplianceUrl /t REG_SZ /d https://conjur.cyberark.com:$(conjur_host_port)/
+        reg ADD HKLM\Software\CyberArk\Conjur /v Version /t REG_DWORD /d 5
+        reg ADD HKLM\Software\CyberArk\Conjur /v Account /t REG_SZ /d cucumber
+        reg ADD HKLM\Software\CyberArk\Conjur /v CertFile /t REG_SZ /d c:\conjur-ca.crt
+
+        # Set conjur.identity equivalent with auth details
+        cmdkey /generic:conjur.cyberark.com /user:${login} /pass:${api_key}
+
+        puppet agent --verbose --onetime --no-daemonize --summarize --masterport $(puppet_host_port) --certname \$(hostname)
+      "
+  set +x
+}
+
+converge_windows_node_hiera_manifest_apikey() {
+  local node_name="hiera-manifest-apikey-node"
+  local hostname="${node_name}_$(openssl rand -hex 3)"
+
+  local login="host/$node_name"
+  local api_key=$(get_host_key $node_name)
+  echo "API key for $node_name: $api_key"
+
+  local hiera_config_file="./code/data/nodes/$hostname.yaml"
+  local manifest_config_file="./code/environments/production/manifests/00_$hostname.pp"
+
+  local ssl_certificate="$(cat https_config/ca.crt | sed 's/^/  /')"
+
+  echo "---
+lookup_options:
+  '^conjur::authn_api_key':
+    convert_to: 'Sensitive'
+
+conjur::account: 'cucumber'
+conjur::appliance_url: 'https://conjur.cyberark.com:$(conjur_host_port)'
+conjur::authn_login: 'host/$node_name'
+conjur::authn_api_key: '$api_key'
+conjur::ssl_certificate: |
+$ssl_certificate
+  " > $hiera_config_file
+
+  echo "
+    node '$hostname' {
+      \$pem_file  = 'c:\test.pem'
+      \$secret = Sensitive(Deferred(conjur::secret, ['inventory/db-password', {
+          appliance_url => lookup('conjur::appliance_url'),
+          account => lookup('conjur::account'),
+          authn_login => lookup('conjur::authn_login'),
+          authn_api_key => lookup('conjur::authn_api_key'),
+          ssl_certificate => lookup('conjur::ssl_certificate')
+      }]))
+
+      notify { \"Writing secret to \${pem_file}...\": }
+      file { \$pem_file:
+        ensure  => file,
+        content => \$secret,
+      }
+
+      exec { \"Read secret from \${pem_file}...\":
+        command => \"C:\Windows\System32\cmd.exe /c type \${pem_file}\",
+        logoutput => true,
+      }
+    }" > $manifest_config_file
+
+  revoke_cert_for "$hostname"
```

Comment:
There doesn't seem to be anything different in this above section between regular and windows setup for hiera. Could we combine the two sections somehow?

Reply:
I don't know a great way to keep track of structured state in bash; there are multiple variables being reused in any one of these functions... I haven't come up with clean way to dry this up just yet. Definitely open to suggestions

Reply:
I guess we can get this into the codebase and see about a cleanup after that